content/blog/tidyr-1-3-0/index.Rmd

```diff
@@ -0,0 +1,289 @@
+    GEOID,
+    widths = c(state = 2, county = 3, tract = 6, block = 4)
+  )
+```
+
+The `name` column contains this same information in a text form, with each component separated by a comma.
+We can use `separate_wider_delim()` to break up this sort of data into individual variables:
+
+```{r}
+vt_census |>
+  select(NAME) |> 
+  separate_wider_delim(
+    NAME,
+    delim = ", ",
+    names = c("block", "block_group", "tract", "county", "state")
+  )
+```
+
+You'll notice that each row contains a lot of duplicated information ("Block", "Block Group", ...).
+You could certainly use `mutate()` and string manipulation to clean this up, but there's a more direct approach that you can use if you're familiar with regular expressions.
+The new `separate_wider_regex()` takes a vector of regular expressions that are matched in order, from left to right.
+If you name the regular expression, it will appear in the output; otherwise, it will be dropped.
+I think this leads to a particularly elegant solution to many problems.
+
+```{r}
+vt_census |>
+  select(NAME) |> 
+  separate_wider_regex(
+    NAME,
+    patterns = c(
+      "Block ", block = "\\d+", ", ",
+      "Block Group ", block_group = "\\d+", ", ",
+      "Census Tract ", tract = "\\d+.\\d+", ", ",
+      county = "[^,]+", ", ",
+      state = ".*"
+    )
+  )
+```
+
+These functions also have a new way to report problems.
+Let's start with a very simple example:
+
+```{r}
+#| error: true
+df <- tibble(
+  id = 1:3,
+  x = c("a", "a-b", "a-b-c")
+)
+
+df |> separate_wider_delim(x, delim = "-", names = c("x", "y"))
+```
+
+We've requested two columns in the output (`x` and `y`), but the first row has only one element and the last row has three elements, so `separate_wider_delim()` can't do what we've asked.
+The error lays out your options for resolving the problem using the `too_few` and `too_many` arguments.
+I'd recommend always starting with `"debug"` to get more information about the problem:
+
+```{r}
+probs <- df |> 
+  separate_wider_delim(
+    x,
+    delim = "-",
+    names = c("a", "b"),
+    too_few = "debug",
+    too_many = "debug"
+  )
+probs
+```
+
+This adds three new variables: `x_ok` tells you if the `x` could be separated as you requested, `x_pieces` tells you the actual number of pieces, and `x_remainder` shows you anything that remains after the columns you asked for.
+You can use this information to fix the problems in the input, or you can use the other options to `too_few` and `too_many` to tell `separate_wider_delim()` to fix them for you:
+
+```{r}
+df |> 
+  separate_wider_delim(
+    x,
+    delim = "-",
+    names = c("a", "b"),
+    too_few = "align_start",
+    too_many = "drop"
+  )
+```
+
+`too_few` and `too_many` also work with `separate_wider_position()`, and `too_few` works with `separate_wider_regex()`.
+The longer variants don't need these arguments because varying numbers of rows don't matter in the same way that varying numbers of columns do:
```

Comment:
```suggestion
The `longer` variants don't need these arguments because varying numbers of rows don't matter in the same way that varying numbers of columns do:
```